// sharyx-voice-agent/dist/index.js
"use strict";
/**
 * Sharyx Voice Agent SDK
 * Build AI voice agents in 5 lines of code.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MockTTS = exports.MockSTT = exports.MockLLM = exports.WebCallAdapter = exports.PlivoAdapter = exports.TwilioAdapter = exports.CartesiaTTS = exports.ElevenLabsTTS = exports.GeminiLLM = exports.OpenAILLM = exports.DeepgramSTT = exports.VoicePipelineService = exports.TelephonyControllers = exports.TelephonyService = exports.Pipeline = exports.VoiceAgent = exports.createAgent = void 0;
// === TIER 1: Simple API (what 90% of people use) ===
var simple_1 = require("./core/simple");
Object.defineProperty(exports, "createAgent", { enumerable: true, get: function () { return simple_1.createAgent; } });
// === TIER 2: Advanced API (power users) ===
var voice_agent_1 = require("./core/voice-agent");
Object.defineProperty(exports, "VoiceAgent", { enumerable: true, get: function () { return voice_agent_1.VoiceAgent; } });
var pipeline_1 = require("./core/pipeline");
Object.defineProperty(exports, "Pipeline", { enumerable: true, get: function () { return pipeline_1.Pipeline; } });
var telephony_system_1 = require("./core/telephony-system");
Object.defineProperty(exports, "TelephonyService", { enumerable: true, get: function () { return telephony_system_1.TelephonyService; } });
Object.defineProperty(exports, "TelephonyControllers", { enumerable: true, get: function () { return telephony_system_1.TelephonyControllers; } });
Object.defineProperty(exports, "VoicePipelineService", { enumerable: true, get: function () { return telephony_system_1.VoicePipelineService; } });
// === Built-in Providers (no extra installs) ===
var deepgram_1 = require("./providers/stt/deepgram");
Object.defineProperty(exports, "DeepgramSTT", { enumerable: true, get: function () { return deepgram_1.DeepgramSTT; } });
var openai_1 = require("./providers/llm/openai");
Object.defineProperty(exports, "OpenAILLM", { enumerable: true, get: function () { return openai_1.OpenAILLM; } });
var gemini_1 = require("./providers/llm/gemini");
Object.defineProperty(exports, "GeminiLLM", { enumerable: true, get: function () { return gemini_1.GeminiLLM; } });
var elevenlabs_1 = require("./providers/tts/elevenlabs");
Object.defineProperty(exports, "ElevenLabsTTS", { enumerable: true, get: function () { return elevenlabs_1.ElevenLabsTTS; } });
var cartesia_1 = require("./providers/tts/cartesia");
Object.defineProperty(exports, "CartesiaTTS", { enumerable: true, get: function () { return cartesia_1.CartesiaTTS; } });
// === Telephony Adapters ===
var twilio_1 = require("./adapters/twilio");
Object.defineProperty(exports, "TwilioAdapter", { enumerable: true, get: function () { return twilio_1.TwilioAdapter; } });
var plivo_1 = require("./adapters/plivo");
Object.defineProperty(exports, "PlivoAdapter", { enumerable: true, get: function () { return plivo_1.PlivoAdapter; } });
var webcall_1 = require("./adapters/webcall");
Object.defineProperty(exports, "WebCallAdapter", { enumerable: true, get: function () { return webcall_1.WebCallAdapter; } });
// === Mocks (for testing) ===
var mock_llm_1 = require("./providers/llm/mock-llm");
Object.defineProperty(exports, "MockLLM", { enumerable: true, get: function () { return mock_llm_1.MockLLM; } });
var mock_stt_1 = require("./providers/stt/mock-stt");
Object.defineProperty(exports, "MockSTT", { enumerable: true, get: function () { return mock_stt_1.MockSTT; } });
var mock_tts_1 = require("./providers/tts/mock-tts");
Object.defineProperty(exports, "MockTTS", { enumerable: true, get: function () { return mock_tts_1.MockTTS; } });
